src/alarms/alarm-handling.md

```diff
@@ -0,0 +1,87 @@
+
+# Information Set per Alarm
+The specified alarm attributes below are inspired by cumulocity data model for alarms, but shall be re-usable for other clouds where possible.
+
+| Name              | Description   |
+| ----------------- |---------------|
+| ref-string        | Device-unique reference string (kind of alarm ID or alarm name), used to reference the once occured alarm again, e.g. in case of updates (e.g. "temperature_sensor_loss") |
+| message-string    | Human readable short information about alarm reason (e.g. "Temperature sensor does not respond") |
+| severity-string   | Could be "CRITICAL", "MAJOR", "MINOR" or "WARNING" |
+| status-string     | Could be "ACTIVE" or "CLEARED"<br/>
+| time-string       | Timestamp indicating when the alarm (or the alarm-update) had occured (in [ISO 8601 format](http://www.w3.org/TR/NOTE-datetime#)) |
+
+# Data flow
+
+Figure below illustrates the data flow from Customer Application broker/thin-edge up to cloud. It shows especially the use of MQTT QOS=1, MQTT Retain and Mosquitto Persistence to achieve reliability alarm state-change transfer.
+
+![Sequence Diagram Update SW-list](images/alarm_dataflow.svg)
+
+**TODO:** Interface from Mapper to Cloud to be defined. Two options are possible:
+1) **JSONviaMQTT:** To have reliability use QOS=1. Open issue: Not yet completely confirmed that C8Y sends PUBACK when msg was processed, instead just on arrival.
+2) **HTTP REST:** HTTP response indicates if message was processed sucessfully. Anyway HTTP REST would increase complexity of implementation and increase data traffic.
+
+
+# Public MQTT-based alarm interface
+
+Similar to thin-edge'S measurement interface, alarm interface is based on MQTT topics.
+
+## Topic structure and payload
+
+Proposal to have "severity" and "ref-string" as topics:
+
+**thin-edge JSON (alarm) format:**
+```
+Topic:   tedge/alarms/<device-id>/<severity-string>/<ref-string>
+Payload: {
+  "message":  <message-string>,
+  "status":   <status-string>,  
+  "time":     <time-string>
+}
+```
+
+NOTE: The "device-id" indicates whether it is an alarm of the thin-edge device it-self (root-device) or a child-device. Use "root" for the thin-edge device or the child-device id for the according child-device.
+
+Benefit to have "severity" and "ref-string" as topics:
+Device-Site reactions to alarm could be easier relized.<br/>
+Examples:
+```
+# Listen to all major alarms on all devices
+tedge mqtt sub "tedge/alarms/+/major/+"
+
+# Listen to specific alarm on all devices
+# (e.g. a bunch of temperature sensors that have each an upper limit setpoint with alarm)
+tedge mqtt sub "tedge/alarms/+/+/upper-limit-exceeded"
+
+# And for sure all combinations.
+```
+
+
+Additionally an also supported "raw"-payload format will allow easy adoption for existing "MQTT publishers":
+
+**Raw payload format:**
+```
+Topic:   tedge/alarms/<device-id>/<severity-string>/<ref-string>
+Payload: <message-string>
+```

```

Comment:
I'm not sure why are we trying to introduce another type of payload, if this payload type is more universal why not commit to use only one?

I think I would prefer this one instead of JSON and having to implement both will add complexity.

Reply:
That "raw payload format" does not contain any timestamp. But I see we shall provide a way to included a timestamp of the alarm occurrence, if the application has any. I don't see a proper way to do that with the "raw payload format".